File: android/discovery/src/main/java/io/github/junekim0007/cryptobench/discovery/PropertyKey.java
package io.github.junekim0007.cryptobench.discovery;

/** Every Provider property-map trap lives here: the Provider.id namespace, aliases, and attributes. */
final class PropertyKey {

    enum Kind { PROVIDER_META, ALIAS, ATTRIBUTE, SERVICE_IMPL, MALFORMED }

    private static final String META_PREFIX = "Provider.";
    private static final String ALIAS_PREFIX = "Alg.Alias.";

    private final Kind kind;
    private final String type;
    private final String algorithm;
    private final String attribute;

    private PropertyKey(Kind kind, String type, String algorithm, String attribute) {
        this.kind = kind;
        this.type = type;
        this.algorithm = algorithm;
        this.attribute = attribute;
    }

    /**
     * Classifies one raw key of the provider property map.
     *
     * <pre>
     * "Provider.id name"          -> PROVIDER_META  (never a service type)
     * "Alg.Alias.Cipher.RC4"      -> ALIAS          type=Cipher, attribute=RC4, value=target
     * "Cipher.AES SupportedModes" -> ATTRIBUTE      type=Cipher, algorithm=AES
     * "Cipher.AES"                -> SERVICE_IMPL
     * </pre>
     */
    public static PropertyKey classify(String key) {
        if (key == null) {
            return malformed();
        }
        if (key.startsWith(META_PREFIX)) {
            return new PropertyKey(Kind.PROVIDER_META, null, null, null);
        }
        if (key.startsWith(ALIAS_PREFIX)) {
            String rest = key.substring(ALIAS_PREFIX.length());
            int dot = rest.indexOf('.');
            if (dot <= 0 || dot == rest.length() - 1) {
                return malformed();
            }
            return new PropertyKey(Kind.ALIAS, rest.substring(0, dot), null, rest.substring(dot + 1));
        }
        int space = key.indexOf(' ');
        if (space < 0) {
            return split(Kind.SERVICE_IMPL, key, null);
        }
        return split(Kind.ATTRIBUTE, key.substring(0, space), key.substring(space + 1).trim());
    }

    private static PropertyKey split(Kind kind, String head, String attribute) {
        int dot = head.indexOf('.');
        if (dot <= 0 || dot == head.length() - 1) {
            return malformed();
        }
        if (kind == Kind.ATTRIBUTE && (attribute == null || attribute.isEmpty())) {
            return malformed();
        }
        return new PropertyKey(kind, head.substring(0, dot), head.substring(dot + 1), attribute);
    }

    private static PropertyKey malformed() {
        return new PropertyKey(Kind.MALFORMED, null, null, null);
    }

    public Kind getKind() {
        return kind;
    }

    public String getType() {
        return type;
    }

    public String getAlgorithm() {
        return algorithm;
    }

    public String getAttribute() {
        return attribute;
    }

    public ServiceKey serviceKey() {
        return ServiceKey.of(type, algorithm);
    }
}

File: android/discovery/src/main/kotlin/probe/PropertyKey.kt
package io.github.junekim0007.cryptobench.discovery.probe

import io.github.junekim0007.cryptobench.discovery.ServiceKey

/**
 * One classified key of a provider property map. A subclass carries only the parts its kind
 * guarantees, so there is no field that is meaningful for some kinds and null for others.
 */
internal sealed class PropertyKey {

    /** `Provider.id name` — never a service type. */
    object ProviderMeta : PropertyKey()

    /** No dot, or an empty part. */
    object Malformed : PropertyKey()

    /** `Alg.Alias.Cipher.RC4` = `ARC4` — [name] is the alias, the property value is the target. */
    data class Alias(val type: String, val name: String) : PropertyKey()

    /** `Cipher.AES SupportedModes` = `ECB|CBC` — [name] is the attribute. */
    data class Attribute(val type: String, val algorithm: String, val name: String) : PropertyKey() {
        val serviceKey: ServiceKey get() = ServiceKey.of(type, algorithm)
    }

    /** `Cipher.AES` — services are read from getServices(), so this is recorded and ignored. */
    data class ServiceImpl(val type: String, val algorithm: String) : PropertyKey()

    companion object {

        private const val META_PREFIX = "Provider."
        private const val ALIAS_PREFIX = "Alg.Alias."

        fun classify(key: String?): PropertyKey {
            if (key == null) {
                return Malformed
            }
            if (key.startsWith(META_PREFIX)) {
                return ProviderMeta
            }
            if (key.startsWith(ALIAS_PREFIX)) {
                val remainder = key.substring(ALIAS_PREFIX.length)
                val separator = remainder.indexOf('.')
                return if (separator <= 0 || separator == remainder.length - 1) {
                    Malformed
                } else {
                    Alias(remainder.substring(0, separator), remainder.substring(separator + 1))
                }
            }
            val space = key.indexOf(' ')
            val serviceSection = if (space < 0) key else key.substring(0, space)
            val separator = serviceSection.indexOf('.')
            if (separator <= 0 || separator == serviceSection.length - 1) {
                return Malformed
            }
            val type = serviceSection.substring(0, separator)
            val algorithm = serviceSection.substring(separator + 1)
            if (space < 0) {
                return ServiceImpl(type, algorithm)
            }
            val attribute = key.substring(space + 1).trim()
            return if (attribute.isEmpty()) Malformed else Attribute(type, algorithm, attribute)
        }
    }
}
